### PR TITLE
Don't throw 404 errors when unarchiving, if the doc is already unarchived

### PR DESCRIPTION
**Pull request by gh2k**
### Description

Unarchiving will throw an error if a doc is already unarchived, which isn't helpful.

It's also possible that there could be a race condition when running this process, if two unarchive processes are triggered at once, so adds a double-check that something else hasn't already unarchived the file while we are processing it..

#### Related Issues / PRs

Should resolve the spurious errors from https://github.com/overleaf/issues/issues/3429 - we should monitor that after deploying to see if any more errors show up.

### Review

There's an indentation change on a bunch of tests because I put them into a new `describe` block, which is a bit annoying. If it's possible to look at the diff without whitespace changes that would be better.

#### Potential Impact

Affects unarchiving, so would prevent opening of an archived project if something goes wrong

#### Manual Testing Performed

- [ ] Archive a project and then open it

### Monitoring

We should watch the logs or metrics for 404s once this is out

### CODE REVIEW

File: app/js/DocArchiveManager.js [hunk 1]
```diff
@@ -93,15 +93,33 @@ async function unarchiveDoc(projectId, docId) {
...
+    stream = await PersistorManager.getObjectStream(
+      settings.docstore.bucket,
+      key
+    )
+  } catch (err) {
+    // if we get a 404, we could be in a race and something else has unarchived the doc already
+    if (err instanceof Errors.NotFoundError) {
+      const doc = await MongoManager.findDoc(projectId, docId, { inS3: 1 })
+      if (!doc.inS3) {
+        // the doc still thinks it is in S3, so that's an error
```
Comment (emcsween) [inline app/js/DocArchiveManager.js:117]: Is this comment meant to describe the other branch (with the `throw`)?

**Review by emcsween — APPROVED** [review]
Just a detail about a comment.

**Review by briangough — APPROVED** [review]

**Comment by briangough** [comment]
Could add some logging for this case (or at least when there's an exception) - it should be rare? 
Archive and unarchive might benefit from some locking in web.

**Comment by gh2k** [comment]
> it should be rare

It happens a few times an hour. I haven't had a dig into why. Could be a bug in the archive job that hits a project twice, but I can't see why it would do that.

**Comment by gh2k** [comment]
My comment above is backwards - this is an unarchiving issue, not an archiving one. We could see this if the project is opened twice at once. I don't think it would be helpful to write out a log item here (the patch prevents this happening) as it shouldn't be OK to do this. - i.e. saying "unarchive" on an unarchived projcet should do nothing and not error, which is usually the case unless you do it two times fast.

**Pull request merged**